cftime/_cftime.pyx

```diff
@@ -121,344 +121,163 @@ def _datesplit(timestr):
-    `calendar='standard'` or `'gregorian'`
-    and the date is after the breakpoint between the Julian and
-    Gregorian calendars (1582-10-15). Otherwise, they are ctime.datetime
-    objects which support some but not all the methods of native python
-    datetime objects. The datetime instances
-    do not contain a time-zone offset, even if the specified `units`
-    contains one.
+    with approximately 1 microsecond accuracy.
     """
     calendar = calendar.lower()
-    basedate = _dateparse(units)
-
-    can_use_python_datetime=\
-      ((calendar == 'proleptic_gregorian' and basedate.year >= MINYEAR) or \
-       (calendar in ['gregorian','standard'] and basedate > gregorian))
-    if not only_use_cftime_datetimes and only_use_python_datetimes:
-        if not can_use_python_datetime:
-            msg='illegal calendar or reference date for python datetime'
-            raise ValueError(msg)
-
-    (unit, ignore) = _datesplit(units)
-
+    basedate = _dateparse(units,calendar=calendar)
+    (unit, isostring) = _datesplit(units)
     # real-world calendars limited to positive reference years.
     if calendar in ['julian', 'standard', 'gregorian', 'proleptic_gregorian']:
         if basedate.year == 0:
             msg='zero not allowed as a reference year, does not exist in Julian or Gregorian calendars'
             raise ValueError(msg)
+    if unit not in UNIT_CONVERSION_FACTORS:
+        raise ValueError("Unsupported time units provided, {!r}.".format(unit))
+    if unit in ["months", "month"] and calendar != "360_day":
+        raise ValueError("Units of months only valid for 360_day calendar.")
+    factor = UNIT_CONVERSION_FACTORS[unit]
+    use_python_datetime = _can_use_python_datetime(basedate,calendar)
 
-    use_python_datetime = False
-    if only_use_python_datetimes and not only_use_cftime_datetimes:
-        # only_use_cftime_datetimes takes precendence
-        use_python_datetime = True
-    if not only_use_python_datetimes and not only_use_cftime_datetimes and can_use_python_datetime:
-        # if only_use_cftimes_datetimes and only_use_python_datetimes are False
-        # return python datetime if possible.
-        use_python_datetime = True
-
-    if use_python_datetime: # use python datetime module
-        isscalar = False
-        try:
-            times[0]
-        except:
-            isscalar = True
-        ismasked = False
-        if np.ma.isMA(times) and np.ma.is_masked(times):
-            mask = times.mask
-            ismasked = True
-        if isscalar:
-            times = np.array([times],dtype='d')
+    isscalar = False
+    try:
+        dates[0]
+    except:
+        isscalar = True
+    ismasked = False
+    if np.ma.isMA(dates) and np.ma.is_masked(dates):
+        mask = dates.mask
+        ismasked = True
+    if isscalar:
+        dates = np.array([dates])
+    else:
+        dates = np.array(dates)
+        shape = dates.shape
+    times = []
+    for date in dates.flat:
```

Comment:
Somewhat related to this, I think there may still be something amiss with respect to masked arrays:
```
In [1]: import numpy as np

In [2]: import cftime

In [3]: times = np.array([cftime.DatetimeNoLeap(2000, 1, 1), cftime.DatetimeNoLeap(2000, 1, 2)])

In [4]: times = np.ma.masked_array(times, mask=[False, True])

In [5]: times
Out[5]:
masked_array(data=[cftime.DatetimeNoLeap(2000, 1, 1, 0, 0, 0), --],
             mask=[False,  True],
       fill_value='?',
            dtype=object)

In [6]: cftime.date2num(times, units="days since 2000-01-01")
Out[6]:
masked_array(data=[0., 1.],
             mask=False,
       fill_value=1e+20)
```
I would expect the last value of the result to remain masked.

Reply:
Thanks for catching that __repr__ bug!  Fixed now, along with the masked array bug.  
Getting rid of the loop in date2num is going to be trickier, since there are so may checks/if tests in there.  Going to leave that for another PR.